### ISSUE TITLE
Upgrade azure-storage-blob to >=12

### ISSUE THREAD

**Issue by cccs-cat001** [apache/airflow#11968]
**Description**
It would be nice if the azure-storage-blob dependency for apache-airflow-providers-microsoft-azure was upgraded to a more recent version that 2.1.0. 

**Use case / motivation**

The version 12.4.0 will allow us to perform OAuth refresh token operations for long running tasks. We use this to authenticate against our datalake, and it's currently not possible with the version 2.1.0 that airflow uses. 

**Related Issues**

The 12.x version is 100% incompatible with the 2.x versions. The API has changed drastically between the two versions, so all the airflow azure operators will need to be changed and fixed.

**Comment by mik-laj** [comment]
https://github.com/apache/airflow/issues/9745 is it related?

**Comment by cccs-cat001** [comment]
I don't think so, that appears to not include any breaking changes, but by upgrading the azure-storage-blob to 12.x the azure WASB hook will break, in favour of the ABFS (I think that's what it's called..).

**Comment by ephraimbuddy** [comment]
We tried this before and met challenges. #8184.

**Comment by cccs-cat001** [comment]
..... Can we try it again? :P

**Comment by ephraimbuddy** [comment]
> ..... Can we try it again? :P

Yeah. I think it will work now. Snowflake now supports azure-storage-blob v12.

**Comment by mik-laj** [comment]
@ephraimbuddy @cccs-cat001 I assigned you to this ticket.

**Comment by ldacey** [comment]
I am using these libraries for Dask and pyarrow (and it is an optional dependency for pandas too). It seems like the adlfs library has a lot of the common methods needed for a wasb_hook:

https://github.com/dask/adlfs/blob/v0.5.0/adlfs/spec.py

https://github.com/intake/filesystem_spec/tree/master/fsspec

I am using a custom wasb_hook right now so I was planning on trying to upgrade to v12 to see what breaks. I am mostly worried about my remote logging (I have not looked into airflow logging at all).

**Comment by ldacey** [comment]
I installed Airflow 1.10.12 and upgraded the Azure Blob SDK to v12 and was able to begin a rough wasb_hook. I still need to add some more methods to upload data, copy blobs, delete blobs, etc. I also need to add the async stuff as well since that is one of the major benefits of v12. Do you think I should just add flags (async=True) with if statements in my current methods where applicable in terms of best practices?

```
import logging
from pathlib import Path

import azure.common
import fsspec
import pendulum
import requests
from airflow.exceptions import AirflowException
from airflow.hooks.base_hook import BaseHook
from azure.storage.blob import BlobServiceClient

logger = logging.getLogger("azure.storage")
logger.setLevel(logging.ERROR)


class AzureBlobHook(BaseHook):
    """
    Interacts with Azure Blob Storage through the wasb:// protocol.
    Additional options passed in the 'extra' field of the connection will be
    passed to the `BlobServiceClient()` constructor. Authenticate through the connection password
    using the account key or a SAS token.
    :param wasb_conn_id: Reference to the wasb connection.
    :type wasb_conn_id: str
    """

    def __init__(self, wasb_conn_id="wasb_default", request_session=None, session_pool_size=10):
        self.conn_id = wasb_conn_id
        self.request_session = request_session
        self.session_pool_size = session_pool_size
        self.conn_info = self.get_connection(self.conn_id)
        self.account_name = self.conn_info.login
        self.connection = self.get_conn()

    def get_conn(self):
        """Return the BlobServiceClient object."""
        service_options = self.conn_info.extra_dejson
        if self.request_session is not None:
            session = requests.Session()
            adapter = requests.adapters.HTTPAdapter(
                pool_connections=self.session_pool_size, pool_maxsize=self.session_pool_size
            )
            session.mount("https://", adapter)
        else:
            session = None
        return BlobServiceClient(
            account_url=f"https://{self.conn_info.login}.blob.core.windows.net",
            credential=self.conn_info.password,
            request_session=session,
            **service_options,
        )

    def list_containers(self, prefix, include_metadata=True, **kwargs):
        """
        Returns a generator to list the containers under the specified account.
        :param prefix: Prefix of the container name.
        :type prefix: str
        :param include_metadata: Whether to include container metadata
        :type include_metadata: bool
        :param kwargs: Optional keyword arguments that
            `BlobServiceClient.list_containers()` takes.
        :type kwargs: object
        """
        return self.connection.list_containers(
            name_starts_with=prefix, include_metadata=include_metadata, **kwargs
        )

    def delete_container(self, container_name, lease=None, **kwargs):
        """
        The container and any blobs contained within it are later deleted during garbage collection.
        If the container is not found, a ResourceNotFoundError will be raised.
        :param container_name: Name of the container
        :type container_name: str
        :param lease: If specified, delete_container only succeeds if the container's lease is active and matches this ID.
                      Required if the container has an active lease.
        :type lease: str
        :param kwargs: Optional keyword arguments that
            `BlobServiceClient.delete_container()` takes.
        :type kwargs: object
        """
        return self.connection.delete_container(container_name, lease=None, **kwargs)

    def create_container(self, container_name, metadata=None, public_access=None, **kwargs):
        """
        Creates a new container under the specified account.
        :param container_name: Name of the container
        :type container_name: str
        :param metadata: A dict with name-value pairs to associate with the container as metadata.
        :type metadata: dict
        :param public_access: Possible values include: 'container', 'blob'.
        :type public_access: str
        :param kwargs: Optional keyword arguments that
            `BlobServiceClient.create_container()` takes.
        :type kwargs: object
        """
        return self.connection.create_container(
            name=container_name, metadata=metadata, public_access=public_access, **kwargs
        )

    def container_client(self, container_name):
        """
        Get a client to interact with the specified container.
        :param container_name: Name of the container
        :type container_name: str
        """
        return self.connection.get_container_client(container=container_name)

    def blob_client(self, container_name, blob_name):
        """
        Get a client to interact with the specified container.
        :param container_name: Name of the container
        :type container_name: str
        :param blob_name: Name of the blob
        :type blob_name: str
        """
        return self.container_client(container_name=container_name).get_blob_client(blob=blob_name)

    def get_bytes(self, container_name, blob_name, start=None, end=None, **kwargs):
        """
        Downloads a blob to the StorageStreamDownloader.
        The readall() method must be used to read all the content or readinto() must be used to download the blob into a stream.
        :param container_name: Name of the container
        :type container_name: str
        :param blob_name: Name of the blob
        :type blob_name: str
        :param start: Start byte position to download blob from
        :type start: int
        :param end: End byte position to download blob from
        :type end: int
        :param kwargs: Optional keyword arguments that
            `ContainerClient.download_blob()` takes.
        :type kwargs: object
        """
        return self.container_client(container_name=container_name).download_blob(
            blob=blob_name, offset=start, length=end, **kwargs
        )

    def get_file(self, container_name, file_path, blob_name, **kwargs):
        """
        Download a file from Azure Blob Storage to the local filesystem.
        :param container_name: Name of the container
        :type container_name: str
        :param blob_name: Name of the blob
        :type blob_name: str
        :param file_path: Destination file path
        :type file_path: str or Path
        :param start: Start byte position to download blob from
        :type start: int
        :param end: End byte position to download blob from
        :type end: int
        :param kwargs: Optional keyword arguments that
            `ContainerClient.download_blob()` takes.
        :type kwargs: object
        """
        file_path = Path(file_path)
        self.log.info(f"Downloading {container_name}/{blob_name} to {str(file_path)}")
        file_path.parent.mkdir(parents=True, exist_ok=True)
        with open(file_path, "wb") as fp:
            blob = self.container_client(container_name=container_name).download_blob(blob=blob_name)
            fp.write(blob.readall())

    def list_blobs(self, container_name, prefix="", include=None, **kwargs):
        """
        Returns a generator to list the blobs under the specified container. 
        The generator will lazily follow the continuation tokens returned by the service.
        :param container_name: Name of the container
        :type container_name: str
        :param prefix: Blob name prefix
        :type prefix: str
        :param include: Options include: 'snapshots', 'metadata', 'uncommittedblobs', 'copy', 'deleted', 'tags'.
        :type include: str
        :param kwargs: Optional keyword arguments that
            `ContainerClient.list_blobs()` takes.
        :type kwargs: object
        """
        return self.container_client(container_name=container_name).list_blobs(
            name_starts_with=prefix, include=include, **kwargs
        )

    def get_blob_properties(self, container_name, blob_name, **kwargs):
        """Returns all user-defined metadata, standard HTTP properties, and system properties for the blob. 
        It does not return the content of the blob.
        :param blob_name: Name of the blob
        :type blob_name: str
        :param kwargs: Optional keyword arguments that
            `BlobClient.get_blob_properties()` takes.
        :type kwargs: object
        """
        return self.blob_client(container_name=container_name, blob_name=blob_name).get_blob_properties(
            **kwargs
        )

    def list_modified_blobs(self, container_name, prefix="", include=None, modified_since_seconds=None):
        """
        Returns a list of blobs modified since a specified period of time.
        :param container_name: Name of the container
        :type container_name: str
        :param prefix: Blob name prefix
        :type prefix: str
        :param include: Options include: 'snapshots', 'metadata', 'uncommittedblobs', 'copy', 'deleted', 'tags'.
        :type include: str
        :param kwargs: Optional keyword arguments that
            `ContainerClient.list_blobs()` takes.
        :type kwargs: object
        """
        blobs = self.list_blobs(container_name=container_name, prefix=prefix, include=include)
        files = []
        for blob in blobs:
            print(blob.name)
            properties = self.get_blob_properties(container_name=container_name, blob_name=blob.name)
            print(properties.last_modified)
            if (
                pendulum.now().int_timestamp - pendulum.parse(str(properties.last_modified)).int_timestamp
                < modified_since_seconds
            ):
                files.append(blob)
        return files

    def create_filesystem(self):
        """Instantiate filesystems for given protocol and arguments"""
        return fsspec.filesystem(
            protocol="abfs", account_name=self.conn_info.login, account_key=self.conn_info.password
        )
```

**Comment by mik-laj** [comment]
> I also need to add the async stuff as well since that is one of the major benefits of v12.

Support for asynchronous operations in Airflow is the future and no operator has yet to support it. For now, I would focus on the synchronous version, and as the next PR we can add asynchronous operations. WDYT?

**Comment by ldacey** [comment]
> > I also need to add the async stuff as well since that is one of the major benefits of v12.
> 
> Support for asynchronous operations in Airflow is the future and no operator has yet to support it. For now, I would focus on the synchronous version, and as the next PR we can add asynchronous operations. WDYT?

I agree with that. Do you think the synchronous v12 support would be in a 1.10 Airflow version or would it have to wait for 2.0?

@ephraimbuddy Any thoughts about keeping the hook method names the same as the previous wasb_hook? I know I personally have a lot of references to the hook in my existing code because I use it for schedule and also for reading/testing random blobs (jupyter notebooks all over the place...). I am perfectly okay with making the changes though, just throwing it out there.

**Comment by ephraimbuddy** [comment]
> > > I also need to add the async stuff as well since that is one of the major benefits of v12.
> > 
> > Support for asynchronous operations in Airflow is the future and no operator has yet to support it. For now, I would focus on the synchronous version, and as the next PR we can add asynchronous operations. WDYT?
> 
> I agree with that. Do you think the synchronous v12 support would be in a 1.10 Airflow version or would it have to wait for 2.0?
> 
> @ephraimbuddy Any thoughts about keeping the hook method names the same as the previous wasb_hook? I know I personally have a lot of references to the hook in my existing code because I use it for schedule and also for reading/testing random blobs (jupyter notebooks all over the place...). I am perfectly okay with making the changes though, just throwing it out there.

Yes. The methods are the same. Extra methods were added too. However, people will need to do the connection differently(I'll still look into this)

**Comment by mik-laj** [comment]
> I agree with that. Do you think the synchronous v12 support would be in a 1.10 Airflow version or would it have to wait for 2.0?

I would like it to be available for Airflow 1.10 and 2.0.

**Comment by karthikBalasubramanian** [comment]
Hello! just wondering if there is any update on this issue :)
